Sources/StreamChat/Controllers/SearchControllers/MessageSearchController/MessageSearchController_Tests.swift

```diff
@@ -0,0 +1,649 @@
+    var controllerCallbackQueueID: UUID!
+    /// Workaround for unwrapping **controllerCallbackQueueID!** in each closure that captures it
+    private var callbackQueueID: UUID { controllerCallbackQueueID }
+    
+    override func setUpWithError() throws {
+        super.setUp()
+        
+        env = TestEnvironment()
+        client = ChatClient.mock
+        query = .init(
+            channelFilter: .exists(.cid),
+            messageFilter: .queryText("")
+        )
+        controller = ChatMessageSearchController(client: client, environment: env.environment)
+        controllerCallbackQueueID = UUID()
+        controller.callbackQueue = .testQueue(withId: controllerCallbackQueueID)
+        // Message search requires a current user
+        client.currentUserId = .unique
+    }
+    
+    override func tearDown() {
+        controllerCallbackQueueID = nil
+        
+        env.messageUpdater?.cleanUp()
+        
+        AssertAsync {
+            Assert.canBeReleased(&controller)
+            Assert.canBeReleased(&client)
+            Assert.canBeReleased(&env)
+        }
+        
+        super.tearDown()
+    }
+    
+    func test_clientIsCorrect() {
+        let controller = client.userSearchController()
+        XCTAssert(controller.client === client)
+    }
+    
+    func test_messagesAreEmpty_beforeSearch() throws {
+        // Save a new message to DB, so DB is not empty
+        try client.databaseContainer.createMessage()
+        
+        // Assert that controller messages is empty
+        XCTAssert(controller.messages.isEmpty)
+    }
+    
+    func test_controllerQueryRemoved_whenControllerIsDeallocated() throws {
+        // Assert that controller messages is empty
+        // Calling `messages` property starts observing DB too
+        XCTAssert(controller.messages.isEmpty)
+        
+        // Make a search
+        controller.search(text: "Hello")
+        
+        // Simulate DB update
+        let messageId = MessageId.unique
+        try client.databaseContainer.createMessage(id: messageId, searchQuery: controller.query)
+        
+        // Simulate network call response
+        env.messageUpdater?.search_completion?(nil)
+        // Release reference of completion so we can deallocate stuff
+        env.messageUpdater?.search_completion = nil
+        
+        var message: ChatMessage? { client.databaseContainer.viewContext.message(id: messageId)?.asModel() }
+        
+        // Check if message is reported
+        AssertAsync.willBeEqual(controller.messages.first, message)
+        
+        let filterHash = controller.query.filterHash
+        // Deallocate controller
+        controller = nil
+        
+        // Assert query doesn't exist in DB anymore
+        AssertAsync.willBeNil(client.databaseContainer.viewContext.messageSearchQuery(filterHash: filterHash))
+        
+        // Assert the message is still here
+        AssertAsync.staysTrue(message != nil)
```

Comment:
If our goal in the future is to move away from `AssertAsync` I think it would be better to stop using it for new tests. Not going to ask you to re-write all these tests ofc, but I think it is worth keeping that in mind for the future.

Reply:
Tests are mostly copy/paste from `UserSearchController_Tests`

Reply:
I see, makes sense then